src/functions.php

```diff
@@ -25,15 +25,29 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
             $loop->cancelTimer($timer);
             $reject($v);
         });
+    }, function ($resolveUnused, $reject) use ($loop, &$timer, $promise) {
+        // cancelling this promise will cancel the timer and reject
+        $loop->cancelTimer($timer);
+        $reject(new \RuntimeException('Timeout cancelled'));
```

Comment:
> The idea for the current cancellation implementation has been to not give consumers the ability to reject a promise through the cancel() call but leave all responsibilities to the promise producer.

I think this makes perfect sense, thanks for the discussion! :+1: 

I've just filed #18 as a better alternative to this PR.